pkg/apis/migration/v1alpha1/migcluster_types.go

```diff
@@ -123,148 +125,66 @@ func (r *MigCluster) DeleteFinalizer() bool {
+// Test the connection settings by building a client.
+func (m *MigCluster) TestConnection(c k8sclient.Client, timeout time.Duration) error {
 	if m.Spec.IsHostCluster {
-		return true, nil
+		return nil
 	}
-
-	// If MigCluster is remote, build a client to check connection
 	restConfig, err := m.BuildRestConfig(c)
 	if err != nil {
-		return false, err
+		return err
 	}
-
-	// Allow setting of custom timeout for connection checking
 	restConfig.Timeout = timeout
-
-	_, err = m.buildClient(restConfig)
+	_, err = k8sclient.New(restConfig, k8sclient.Options{Scheme: scheme.Scheme})
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	return true, nil
+	return nil
 }
 
-// BuildRestConfig creates a remote cluster RestConfig from a MigCluster and a local client
+// Build a REST configuration.
 func (m *MigCluster) BuildRestConfig(c k8sclient.Client) (*rest.Config, error) {
-	// Get first K8s endpoint from ClusterRef
-	clusterRef := m.Spec.ClusterRef
-	cluster := &crapi.Cluster{}
-	if clusterRef == nil {
-		err := errors.New("ClusterRef is nil.")
-		return nil, err
-	}
-
-	err := c.Get(context.TODO(), types.NamespacedName{Name: clusterRef.Name, Namespace: clusterRef.Namespace}, cluster)
+	secret, err := GetSecret(c, m.Spec.ServiceAccountSecretRef)
 	if err != nil {
-		return nil, err // TODO: introspect error type
-	}
-
-	if cluster.Spec.KubernetesAPIEndpoints.ServerEndpoints == nil {
-		return nil, fmt.Errorf("MigCluster [%s/%s] references Cluster [%s/%s] with nil ServerEndpoints",
-			m.Namespace, m.Name, clusterRef.Namespace, clusterRef.Name)
-	}
-
-	k8sEndpoints := cluster.Spec.KubernetesAPIEndpoints.ServerEndpoints
-	if len(k8sEndpoints) < 1 {
-		return nil, fmt.Errorf("MigCluster [%s/%s] references Cluster [%s/%s] with 0 ServerEndpoints",
-			m.Namespace, m.Name, clusterRef.Namespace, clusterRef.Name)
-	}
-
-	if k8sEndpoints[0].ServerAddress == "" {
-		return nil, fmt.Errorf("MigCluster [%s/%s] references Cluster [%s/%s] with an empty ServerAddress at ServerEndpoints[0]",
-			m.Namespace, m.Name, clusterRef.Namespace, clusterRef.Name)
-	}
-	clusterURL := string(k8sEndpoints[0].ServerAddress)
-
-	// Get SA token attached to this MigCluster
-	saSecretRef := m.Spec.ServiceAccountSecretRef
-	saSecret := &kapi.Secret{}
-	if saSecretRef == nil {
-		err := errors.New("ServiceAccountSecretRef not set.")
 		return nil, err
 	}
-
-	err = c.Get(context.TODO(), types.NamespacedName{Name: saSecretRef.Name, Namespace: saSecretRef.Namespace}, saSecret)
-	if err != nil {
-		return nil, err // TODO: introspect error type
-	}
-
-	saTokenKey := "saToken"
-	saTokenData, ok := saSecret.Data[saTokenKey]
-	if !ok {
-		return nil, fmt.Errorf("MigCluster [%s/%s] references SA token secret [%s/%s] with empty 'saToken' field",
-			m.Namespace, m.Name, saSecretRef.Namespace, saSecretRef.Name)
+	restConfig := &rest.Config{
+		Host:        m.Spec.URL,
+		BearerToken: string(secret.Data[SaToken]),
+		TLSClientConfig: rest.TLSClientConfig{
```

Comment:
I see you figured out how to get `Insecure: true` set on the initial object. I was struggling with that.